lxd/auth/driver_openfga.go

```diff
@@ -0,0 +1,933 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/openfga/go-sdk"
+	"github.com/openfga/go-sdk/client"
+	"github.com/openfga/go-sdk/credentials"
+
+	"github.com/canonical/lxd/lxd/request"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/logger"
+)
+
+type fga struct {
+	commonAuthorizer
+
+	apiURL   string
+	apiToken string
+	storeID  string
+
+	client      *client.OpenFgaClient
+	authModelID string
+}
+
+func (f *fga) validateConfig() error {
+	val, ok := f.config["openfga.api.token"]
+	if !ok {
+		return fmt.Errorf("Missing OpenFGA API token")
+	}
+
+	f.apiToken = val.(string)
+
+	val, ok = f.config["openfga.api.url"]
+	if !ok {
+		return fmt.Errorf("Missing OpenFGA API URL")
+	}
+
+	f.apiURL = val.(string)
+
+	val, ok = f.config["openfga.store.id"]
+	if !ok {
+		return fmt.Errorf("Missing OpenFGA store ID")
+	}
+
+	f.storeID = val.(string)
+
+	val, ok = f.config["openfga.store.model_id"]
+	if ok {
+		f.authModelID = val.(string)
+	}
+
+	return nil
+}
+
+func (f *fga) load() error {
+	err := f.validateConfig()
+	if err != nil {
+		return err
+	}
+
+	u, err := url.Parse(f.apiURL)
+	if err != nil {
+		return fmt.Errorf("Failed parsing URL: %w", err)
+	}
+
+	conf := client.ClientConfiguration{
+		ApiScheme: u.Scheme,
+		ApiHost:   u.Host,
+		StoreId:   f.storeID,
+		Credentials: &credentials.Credentials{
+			Method: credentials.CredentialsMethodApiToken,
+			Config: &credentials.Config{
+				ApiToken: f.apiToken,
+			},
+		},
+	}
+
+	f.client, err = client.NewSdkClient(&conf)
+	if err != nil {
+		return fmt.Errorf("Failed to create OpenFGA client: %w", err)
+	}
+
+	var builtinAuthorizationModel client.ClientWriteAuthorizationModelRequest
+
+	err = json.Unmarshal([]byte(authModel), &builtinAuthorizationModel)
```

Comment:
### Migrations
Here `authModel` is the json string that is generated via `make update-openfga`. If there is an existing model in the store, we ensure that this model is exactly equal to the model we have statically defined. This is where we would have an opportunity to detect different model versions and apply update logic.

Migrations in OpenFGA follow a pretty basic idea. The idea is to create an intermediary model which supports both the new and the old, then write all the tuples required to support the new schema, delete all the tuples that are no longer required, and finally write the updated model (see https://openfga.dev/docs/modeling/migrating/migrating-relations).

In practice, I think we will likely have something like:
* Under an `lxd/auth/openfga` directory:
    - `model_v1.openfga`
    - `model_v1.go`
    - `model_v2_compatible.openfga`
    - `model_v2_compatible.go`
    - `model_v2.openfga`
    - `model_v2.go`
    - `migrations.go`
    - `synchronise.go`
* The `migrations.go` file will contain a map of methods which accept an openfga client and performs the necessary actions to migrate from one schema to the next. These methods will only be valid while the compatible model is in place.
* The `synchronise.go` file will contain a map of methods which accept an fga client and a set of LXD resources and ensure they match 1:1 with what is present in the OpenFGA store. Like migrations, this is also model specific.
* When we connect to the OpenFGA server:
    1. Get the current auth model.
    2. Check if it is the latest version. If it is, sync resources and continue. Otherwise....
    3. Write the compatible model.
    4. Call the associated method from `migrations.go`.
    5. Write the latest model.
    6. Perform a final resource sync from `synchronise.go`.

Hopefully we won't have to do this very often.

Reply:
I think we can handle this when we come to update the model.